scheduler/planner.go

```diff
@@ -0,0 +1,348 @@
+	for id := range unit.tasks {
+		ids = append(ids, id)
+	}
+	sort.Sort(ids)
+
+	for _, id := range ids {
+		_, _ = io.WriteString(hash, id)
+	}
+
+	unit.id = fmt.Sprintf("%x", hash.Sum(nil))
+	return unit.id
+}
+
+// RankValue returns a point value for the tasks in the unit that can
+// be used to compare units with eachother.
+//
+// Generally, higher point values are given to larger units and for
+// units that have been in the queue for longer, with longer expected
+// runtimes. The tasks priority act as a multiplying factor.
+func (unit *Unit) RankValue() int64 {
+	if unit.cachedValue > 0 {
+		return unit.cachedValue
+	}
+
+	var (
+		expectedRuntime time.Duration
+		timeInQueue     time.Duration
+		totalPriority   int64
+		numDeps         int64
+		inCommitQueue   bool
+		inPatch         bool
+	)
+
+	for _, t := range unit.tasks {
+		if t.Requester == evergreen.MergeTestRequester {
+			inCommitQueue = true
+		} else if evergreen.IsPatchRequester(t.Requester) {
+			inPatch = true
+		}
+
+		if !t.ScheduledTime.IsZero() {
+			timeInQueue += time.Since(t.ScheduledTime)
+		}
+
+		totalPriority += t.Priority
+		expectedRuntime += t.FetchExpectedDuration()
+		numDeps += int64(t.NumDependents)
+	}
+
+	num := int64(len(unit.tasks))
+	priority := 1 + (totalPriority / num)
+
+	if inCommitQueue {
+		priority += 100
+	}
+
+	if inPatch {
+		unit.cachedValue += unit.distro.GetPatchZipperFactor()
+	}
+
+	unit.cachedValue += num
+	unit.cachedValue += priority
+	unit.cachedValue += priority * (numDeps / num)
+	unit.cachedValue += priority * unit.distro.GetExpectedRuntimeFactor() * int64(math.Floor(expectedRuntime.Minutes()/float64(num)))
+	unit.cachedValue += priority * unit.distro.GetTimeInQueueFactor() * int64(math.Floor(timeInQueue.Minutes()/float64(num)))
+	return unit.cachedValue
+}
+
+// StringSet provides simple tools for managing sets of strings.
+type StringSet map[string]struct{}
+
+// Add places the string in the set.
+func (s StringSet) Add(id string) { s[id] = struct{}{} }
+
+// Check returns true if the string is a member of the set.
+func (s StringSet) Check(id string) bool { _, ok := s[id]; return ok }
+
+// Visit returns true if the string is already a member of the
+// set. Otherwise it adds it to the set and returns false.
+func (s StringSet) Visit(id string) bool {
+	if s.Check(id) {
+		return true
+	}
+
+	s.Add(id)
+	return false
+}
+
+// TaskList implements sort.Interface on top of a slice of tasks. The
+// provided sorting, orders members of task groups, and then
+// prioritizes tasks by the number of dependencies, priority, and
+// expected duration.
+type TaskList []task.Task
+
+func (tl TaskList) Len() int      { return len(tl) }
+func (tl TaskList) Swap(i, j int) { tl[i], tl[j] = tl[j], tl[i] }
+func (tl TaskList) Less(i, j int) bool {
+	t1 := tl[i]
+	t2 := tl[j]
+
+	if t1.TaskGroupOrder != t2.TaskGroupOrder {
+		return t1.TaskGroupOrder < t2.TaskGroupOrder
+	}
+
+	if t1.NumDependents != t2.NumDependents {
```

Comment:
Shouldn't number of dependents be handled by the priority arithmetic? Same question about Priority and expected duration.